borg/archive.py

```diff
@@ -467,23 +467,40 @@ def rename(self, name):
         self.cache.chunk_decref(self.id, self.stats)
         del self.manifest.archives[self.name]
 
-    def delete(self, stats, progress=False):
-        unpacker = msgpack.Unpacker(use_list=False)
-        items_ids = self.metadata[b'items']
-        pi = ProgressIndicatorPercent(total=len(items_ids), msg="Decrementing references %3.0f%%", same_line=True)
-        for (i, (items_id, data)) in enumerate(zip(items_ids, self.repository.get_many(items_ids))):
+    def delete(self, stats, progress=False, forced=False):
+        error = False
+        try:
+            unpacker = msgpack.Unpacker(use_list=False)
+            items_ids = self.metadata[b'items']
+            pi = ProgressIndicatorPercent(total=len(items_ids), msg="Decrementing references %3.0f%%", same_line=True)
+            for (i, (items_id, data)) in enumerate(zip(items_ids, self.repository.get_many(items_ids))):
+                if progress:
+                    pi.show(i)
+                unpacker.feed(self.key.decrypt(items_id, data))
+                self.cache.chunk_decref(items_id, stats)
+                for item in unpacker:
+                    if b'chunks' in item:
+                        for chunk_id, size, csize in item[b'chunks']:
+                            self.cache.chunk_decref(chunk_id, stats)
             if progress:
-                pi.show(i)
-            unpacker.feed(self.key.decrypt(items_id, data))
-            self.cache.chunk_decref(items_id, stats)
-            for item in unpacker:
-                if b'chunks' in item:
-                    for chunk_id, size, csize in item[b'chunks']:
-                        self.cache.chunk_decref(chunk_id, stats)
-        if progress:
-            pi.finish()
-        self.cache.chunk_decref(self.id, stats)
+                pi.finish()
+        except Exception:
+            if not forced:
+                raise
+            error = True
```

Comment:
- Depends on what caused the corruption, but likely.
- Not necessarily. If chunks are missing from the repository then this approach may leave large amounts of unused data (think first few chunks of a large file like a disk image go missing) that wouldn't be removed otherwise
- True, on the other hand over-zealous exception handling led to other bugs in the past.

Also I think it'd be better if the error handling would be more detailed (without tracebacks that can't point to the problem here anyway). If the items stream is corrupted it should tell that the archive metadata was corrupted so the data couldn't be deleted (which seems to be a common case due to #1138). If objects are absent from the Cache that should be fatal, because the cache is corrupted. If objects are absent from the repository that shouldn't be fatal, but it should say that data was already missing.